### ISSUE TITLE
Test: When testing KCP remediation Should replace unhealthy machines is failing

### ISSUE THREAD

**Issue by Sunnatillo** [kubernetes-sigs/cluster-api#10885]
### Which jobs are failing?

capi-e2e-main capi-e2e-mink8s-main

```
capi-e2e: [It] When testing KCP remediation Should replace unhealthy machines expand_less	5m0s
{Timed out after 300.001s.
Timed out waiting for Cluster kcp-remediation-yqc57p/kcp-remediation-5y5a3d to provision
Expected
    <string>: Provisioning
to equal
    <string>: Provisioned failed [FAILED] Timed out after 300.001s.
Timed out waiting for Cluster kcp-remediation-yqc57p/kcp-remediation-5y5a3d to provision
Expected
    <string>: Provisioning
to equal
    <string>: Provisioned
In [It] at: /home/prow/go/src/sigs.k8s.io/cluster-api/test/framework/cluster_helpers.go:144 @ 07/17/24 07:32:51.824

There were additional failures detected after the initial failure. These are visible in the timeline
}
```

```
INFO: Waiting for the cluster infrastructure to be provisioned
  STEP: Waiting for cluster to enter the provisioned phase @ 07/17/24 06:51:19.97
  [FAILED] in [It] - /home/prow/go/src/sigs.k8s.io/cluster-api/test/framework/cluster_helpers.go:144 @ 07/17/24 06:56:19.972
  [PANICKED] in [AfterEach] - /usr/local/go/src/runtime/panic.go:261 @ 07/17/24 06:56:19.973
  << Timeline
  [FAILED] Timed out after 300.001s.
  Timed out waiting for Cluster kcp-remediation-1b56mw/kcp-remediation-k2optk to provision
  Expected
      <string>: Provisioning
  to equal
      <string>: Provisioned
  In [It] at: /home/prow/go/src/sigs.k8s.io/cluster-api/test/framework/cluster_helpers.go:144 @ 07/17/24 06:56:19.972
  Full Stack Trace

```

### Which tests are failing?

 When testing KCP remediation Should replace unhealthy machines is failing

### Since when has it been failing?

16.07  20:04 UTC

### Testgrid link

https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/periodic-cluster-api-e2e-main/1813470576841330688

### Label(s) to be applied

/kind failing-test
/area ci

**Comment by chrischdi** [comment]
Maybe related to:

- #10873

**Comment by sbueringer** [comment]
cc @vincepri

**Comment by Sunnatillo** [comment]
Docker machine not became ready:

`failed to exec DockerMachine bootstrap: failed to run cloud config: stdout: Waiting for signal...\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal`

**Comment by Sunnatillo** [comment]
Docker machine is not becoming ready:
`{"ts":1721221506274.7917,"caller":"controller/controller.go:324","msg":"Reconciler error","controller":"dockermachine","controllerGroup":"infrastructure.cluster.x-k8s.io","controllerKind":"DockerMachine","DockerMachine":{"name":"kcp-remediation-t802hp-control-plane-xlczm","namespace":"kcp-remediation-sixb0a"},"namespace":"kcp-remediation-sixb0a","name":"kcp-remediation-t802hp-control-plane-xlczm","reconcileID":"6ca775c3-8a8b-47da-aae8-b50ccecd3dea","err":"failed to exec DockerMachine bootstrap: failed to run cloud config: stdout: Waiting for signal...\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsignal hold\nsig`

**Comment by sbueringer** [comment]
Did a quick revert PR to verify it was introduced by https://github.com/kubernetes-sigs/cluster-api/pull/10873

(see: https://github.com/kubernetes-sigs/cluster-api/pull/10898)

**Comment by Sunnatillo** [comment]
This line is the problem. In kcp remediation test controlplane does not become ready when workload cluster is initialized. 
https://github.com/kubernetes-sigs/cluster-api/blob/ff54cbff5c72de1aa486b09e4e866b37637c59bd/internal/controllers/cluster/cluster_controller_phases.go#L56

**Comment by Sunnatillo** [comment]
```
if (cluster.Spec.InfrastructureRef == nil || cluster.Status.InfrastructureReady) &&
		(cluster.Spec.ControlPlaneRef == nil || cluster.Status.ControlPlaneReady) &&
		cluster.Spec.ControlPlaneEndpoint.IsValid()
```
(cluster.Spec.ControlPlaneRef == nil || cluster.Status.ControlPlaneReady) is problematic because in kcp remediation test controlplane never becomes ready  and it should start remediation. With that line cluster never becomes provisioned and remediation will not start.

**Comment by Sunnatillo** [comment]
@fabriziopandini 
We can merge revert PR to make CI  green.

**Comment by fabriziopandini** [comment]
I don't think it is correct to revert entirely https://github.com/kubernetes-sigs/cluster-api/pull/10873, because cluster phases must be improved to account for the use case where infrastructure ref is not provided

I have created https://github.com/kubernetes-sigs/cluster-api/pull/10903 to make a small fix that should bring tests back to green
